modules/common/src/main/paginator/Paginator.scala

```diff
@@ -91,11 +91,17 @@ object Paginator {
       currentPage: Int = 1,
       maxPerPage: MaxPerPage = MaxPerPage(10)
   )(implicit ec: scala.concurrent.ExecutionContext): Validated[String, Fu[Paginator[A]]] =
-    if (currentPage < 1) Validated.invalid("Max per page must be greater than zero")
-    else if (maxPerPage.value <= 0) Validated.invalid("Current page must be greater than zero")
+    if (currentPage < 1) Validated.invalid("Current page must be greater than zero")
+    else if (maxPerPage.value <= 0) Validated.invalid("Max per page must be greater than zero")
     else
-      Validated.valid(for {
-        results   <- adapter.slice((currentPage - 1) * maxPerPage.value, maxPerPage.value)
-        nbResults <- adapter.nbResults
-      } yield new Paginator(currentPage, maxPerPage, results, nbResults))
+      Validated.valid(
+        adapter.nbResults flatMap { nbResults =>
+          val safePage = 1 max (currentPage min ((nbResults + maxPerPage.value - 1) / maxPerPage.value))
```

Comment:
instead of min and max it's imo much clearer to use the aliases atLeast and atMost, i.e. maybe:
```scala
val safePage = currentPage atLeast 1 atMost ((nbResults + maxPerPage.value - 1) / maxPerPage.value)
```

not sure if it works like that without parenthesis but you get the idea